src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs

```diff
@@ -72,68 +78,90 @@ public bool TryGenerateBashScript(BuildScriptGeneratorContext context, out strin
         {
-            bool providedLanguageFound = false;
             var resultPlatforms = new List<Tuple<IProgrammingPlatform, string>>();
 
-            foreach (var platform in _programmingPlatforms)
+            var enabledPlatforms = _programmingPlatforms.Where(p =>
             {
-                if (!platform.IsEnabled(ctx))
+                if (!p.IsEnabled(ctx))
                 {
-                    _logger.LogDebug("{platformName} has been disabled", platform.Name);
-                    continue;
+                    _logger.LogDebug("{platformName} has been disabled", p.Name);
+                    return false;
                 }
+                return true;
+            });
 
-                bool usePlatform = false;
-                var currPlatformMatchesProvided = !string.IsNullOrEmpty(ctx.Language) &&
-                    string.Equals(ctx.Language, platform.Name, StringComparison.OrdinalIgnoreCase);
+            // If a user supplied the language explicitly, check if the platform is enabled for that
+            IProgrammingPlatform userSuppliedPlatform = null;
+            string platformVersion = null;
+            if (!string.IsNullOrEmpty(ctx.Language))
+            {
+                var selectedPlatform = enabledPlatforms
+                    .Where(p => string.Equals(ctx.Language, p.Name, StringComparison.OrdinalIgnoreCase))
+                    .FirstOrDefault();
 
-                string targetVersionSpec = null;
-                if (currPlatformMatchesProvided)
+                if (selectedPlatform == null)
                 {
-                    providedLanguageFound = true;
-                    targetVersionSpec = ctx.LanguageVersion;
-                    usePlatform = true;
-                }
-                else if (ctx.DisableMultiPlatformBuild && !string.IsNullOrEmpty(ctx.Language))
-                {
-                    _logger.LogDebug(
-                        "Multi platform build is disabled and platform was specified. " +
-                        "Skipping language {skippedLang}",
-                        platform.Name);
-                    continue;
+                    ThrowInvalidLanguageProvided(ctx);
                 }
 
-                if (!currPlatformMatchesProvided || string.IsNullOrEmpty(targetVersionSpec))
+                userSuppliedPlatform = selectedPlatform;
+
+                platformVersion = ctx.LanguageVersion;
+                if (string.IsNullOrEmpty(platformVersion))
                 {
-                    _logger.LogDebug("Detecting platform using {platformName}", platform.Name);
-                    var detectionResult = platform.Detect(ctx.SourceRepo);
-                    if (detectionResult != null)
+                    var detectionResult = userSuppliedPlatform.Detect(ctx.SourceRepo);
+                    if (string.IsNullOrEmpty(detectionResult.LanguageVersion))
```

Comment:
we should probably check for null here. for, say, an empty repo, we won't detect the language or version, but the user might have specified the language version... this also brings the point that we probably want a test for an empty repo, and agree of what we should do in that case (e.g. fail gracefully at least)

Reply:
There was already a test called `TryGenerateScript_Throws_IfLanguageIsProvidedButNoVersion_AndCannotDetectVersion`, but added a new one for language too: `TryGenerateScript_Throws_IfLanguageIsProvided_AndCannotDetectLanguage`

descriptive name are good ;-)